drivers/watchdog/wdt_xilinx_axi.c

```diff
@@ -0,0 +1,234 @@
+
+enum xilinx_wdt_axi_register {
+	REG_TWCSR0 = 0x00, /* Control/Status Register 0 */
+	REG_TWCSR1 = 0x04, /* Control/Status Register 1 */
+	REG_TBR = 0x08,	   /* Timebase Register */
+	REG_MWR = 0x0C,	   /* Master Write Control Register */
+};
+
+enum xilinx_wdt_csr0_bits {
+	CSR0_WRS = BIT(3),
+	CSR0_WDS = BIT(2),
+	CSR0_EWDT1 = BIT(1),
+	CSR0_EWDT2 = BIT(0),
+};
+
+enum xilinx_wdt_csr1_bits {
+	CSR1_EWDT2 = BIT(0),
+};
+
+enum {
+	TIMER_WIDTH_MIN = 8,
+};
+
+LOG_MODULE_REGISTER(wdt_xilinx_axi, CONFIG_WDT_LOG_LEVEL);
+
+struct xilinx_wdt_axi_config {
+	mem_addr_t base;
+	uint32_t clock_rate;
+	uint8_t timer_width_max;
+	bool enable_once;
+};
+
+struct xilinx_wdt_axi_data {
+	struct k_spinlock lock;
+	bool last_reset_by_wdt;
+	bool timeout_active;
+	bool wdt_started;
+};
+
+static int wdt_xilinx_axi_setup(const struct device *dev, uint8_t options)
+{
+	const struct xilinx_wdt_axi_config *config = dev->config;
+	struct xilinx_wdt_axi_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	int ret;
+
+	if (!data->timeout_active) {
+		ret = -EINVAL;
+		goto out;
+	}
+	if (data->wdt_started) {
+		ret = -EBUSY;
+		goto out;
+	}
+
+	/* We don't actually know or control at the driver level whether
+	 * the WDT pauses in CPU sleep or when halted by the debugger,
+	 * so we don't check anything with the options.
+	 */
+	sys_write32(CSR0_EWDT1 | CSR0_WRS | CSR0_WDS, config->base + REG_TWCSR0);
+	sys_write32(CSR1_EWDT2, config->base + REG_TWCSR1);
+	data->wdt_started = true;
+	ret = 0;
+
+out:
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int wdt_xilinx_axi_disable(const struct device *dev)
+{
+	const struct xilinx_wdt_axi_config *config = dev->config;
+	struct xilinx_wdt_axi_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	int ret;
+
+	if (config->enable_once) {
+		ret = -EPERM;
+		goto out;
+	}
+	if (!data->wdt_started) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	sys_write32(CSR0_WRS | CSR0_WDS, config->base + REG_TWCSR0);
+	sys_write32(0, config->base + REG_TWCSR1);
+	data->wdt_started = false;
+	ret = 0;
+
+out:
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int wdt_xilinx_axi_install_timeout(const struct device *dev,
+					  const struct wdt_timeout_cfg *cfg)
+{
+	const struct xilinx_wdt_axi_config *config = dev->config;
+	struct xilinx_wdt_axi_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	uint32_t timer_width;
+	bool good_timer_width = false;
+	int ret;
+
+	if (data->timeout_active) {
+		ret = -ENOMEM;
+		goto out;
+	}
+	if (!(cfg->flags & WDT_FLAG_RESET_SOC)) {
+		ret = -ENOTSUP;
+		goto out;
+	}
+	if (cfg->window.min != 0) {
+		ret = -EINVAL;
+		goto out;
+	}
+	for (timer_width = TIMER_WIDTH_MIN; timer_width <= config->timer_width_max; timer_width++) {
+		/* Note: WDT expiry happens after 2 wraps of the timer (first raises an interrupt
+		 * which is not used, second triggers a reset) so add 1 to timer_width.
+		 */
```

Comment:
Why not use the first IRQ to trigger the user callback?

Reply:
I looked at this, but it would make things a fair bit more complex to have this really work properly. We didn't need that for our use case and it looks like that behavior wasn't implemented in the Linux driver for this core either.